### DISCUSSION TITLE
Multiple id3v2 TXXX tags for ReplayGain when using TaggedFile

### DISCUSSION THREAD

**Discussion by Danacus** [Serial-ATA/lofty-rs#222] [Q&A] [answered]
I am trying to extract ReplayGain tags from files, as described on the [ReplayGain Specification](https://wiki.hydrogenaud.io/index.php?title=ReplayGain_2.0_specification#ID3v2). For id3v2, All different types of ReplayGain information are stored in separate TXXX tags and the description serves as the key (e.g. `REPLAYGAIN_TRACK_GAIN` or `REPLAYGAIN_ALBUM_GAIN`). However, when using `Tag` in `lofty`, this description seems to get removed, making it impossible to distinguish between different ReplayGain tags, as illustrated below when using `Tag::get_items`;

```
    TagItem {
        item_key: Unknown(
            "TXXX",
        ),
        item_value: Text(
            "-10.43 dB\0",
        ),
    },
    TagItem {
        item_key: Unknown(
            "TXXX",
        ),
        item_value: Text(
            "-10.40 dB\0",
        ),
    },
```

I don't know which of those is album gain and which is track gain, because the description is not stored (from what I can tell).

The [documentation of lofty](https://docs.rs/lofty/latest/lofty/id3/v2/struct.ID3v2Tag.html) mentions that those `TXXX` tags are identified by their descriptions, but I'm not sure how those descriptions can be used. Am I missing something, or is it currently not possible to get the descriptions of `TXXX` when using the generic `Tag` (or a `Tag` converted to a `ID3v2Tag`)?

**Comment by Serial-ATA** [comment]
That's an issue, TXXX/WXXX frames should be removed during the conversion to `Tag`, so thanks! The issue with those is there's no real way for a `TagItem` to store that extra description. What you'll have to do is read a concrete file type (e.g. `MP3File`) and obtain an `ID3v2Tag`.

**Comment by Serial-ATA** [comment]
In the next release, I'll add some `ItemKey`s for ReplayGain so you can use `Tag`, and the conversion isn't lossy.

**Comment by Danacus** [comment]
That would be great, thanks a lot! For now, I'll use a concrete file type, as you suggested. 

I'm currently trying to add support for ReplayGain to amberol, a music player (see https://gitlab.gnome.org/World/amberol/-/issues/75). The maintainer of amberol will most likely prefer a solution using `Tag`, since that's what he's already using right now.

**Comment by Serial-ATA** [comment]
This was simpler to implement than I thought, could try out https://github.com/Serial-ATA/lofty-rs/commit/97f081fb4810695498da925085e7dd752ba074dc? If all goes well I can publish 0.7.3 :slightly_smiling_face:.

**Comment by Danacus** [comment]
That was really fast! I'll try it tomorrow.

**Comment by Danacus** [comment]
This is working well for me. Thanks!

**Comment by Serial-ATA** [comment, marked as answer]
Great, thanks! `0.7.3` is out now :slightly_smiling_face:.
